### ISSUE TITLE
`strictNullChecks` short-circuits the subsequent type checks

### ISSUE THREAD

**Issue by PPpro** [microsoft/TypeScript#53162]
# Bug Report

hello， I encounter a strange bug when I use `strictNullChecks: true`, and here is an example:

```ts
interface A {
  a: number;
}
const list: A[] = [];
list[0] ??= {};  // throw an error: Property 'a' is missing in type '{}' but required in type 'A'
list[0] = list[0] ?? {};  // the type checking not work !
```

### 🕗 Version & Regression Information

- This is a crash
- This is the behavior in every version I tried
### ⏯ Playground Link

It seems the type checking works after I close the `strictNullChecks` options
[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.2.3#code/JYOwLgpgTgZghgYwgAgILIN4FgBQzlwBcyIArgLYBG0A3LgL64ID2IAzmMgDbAfGoBtALrIAvMmF0cPDgIAMIgPyLxGelJlh5I8Zu3JlmekA)

### 💻 Code

```ts
interface A {
  a: number;
}
const list: A[] = [];
list[0] ??= {};  // throw an error: Property 'a' is missing in type '{}' but required in type 'A'
list[0] = list[0] ?? {};  // the type checking not work !
```

### 🙁 Actual behavior

`list[0] = list[0] ?? {};` doens't throw any errors

### 🙂 Expected behavior

I expect that `list[0] = list[0] ?? {}; ` should throw an error

**Comment by jakebailey** [comment]
Any value which implements the interface `A` must be truthy, so it doesn't make sense to emit an error here unless you have `noUncheckedIndexedAccess` enabled, in which case you do see an error: [Playground Link](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&ts=4.9.5#code/JYOwLgpgTgZghgYwgAgILIN4FgBQzlwBcyIArgLYBG0A3LgL64ID2IAzmMgDbAfGoBtALrIAvMmF0cPDgIAMIgPyLxGelJlh5I8WACeABwjMY3XloVjR4gOSkQAEwgxQEBzeSLM9ZMU3aNc20xM1lLZW9AsJ1QixEAH3jIoA)

**Comment by PPpro** [comment]
@jakebailey 
thanks for reply, 
then one thing bothers me that should the statement `list[0] ??= {};` emit an error ?
should we keep the behavior the same as `list[0] = list[0] ?? {};`

**Comment by fatcerberus** [comment]
@jakebailey It does seem inconsistent though that `list[0] ??= {}` produces an error, since the RHS is equally irrelevant by the same logic.

**Comment by jakebailey** [comment]
Ah. I see what you mean.

**Comment by Andarist** [comment]
Re last one: `checkConditionalExpression` produces a union of both true/false types:
https://github.dev/microsoft/TypeScript/blob/746a6feb2e7ba6987b6c72db538dd498b35cd461/src/compiler/checker.ts#L37036

Binary expressions take into account control flow analysis. So in this case, it's known that `|| {}` and `?? {}` can't be chosen by those expressions and thus their type is simply `true` and thus assignable to `a`.

I don't know why conditional expressions don't apply the same control flow-based "narrowing" though.

Re first one: it's likely a weird (at first glance) byproduct of the same control flow analysis. In all of those cases TypeScript checks if the RHS is assignable to the left side. It just happens that RHS type isn't the same in all of them because the variants with `|| {}` and `?? {}` are known to be of type `A` since the first operand of those is truthy and CFA knows that the second operand can't be selected (according to the TS rules around indexed access on arrays). In the `??=` the RHS is different - it's actually `{}` and that type is validated against the assignment target.

**Comment by fatcerberus** [comment]
> In this case, it seems that we should close `strictNullChecks` to make it more type-safe.

I assume by "close" you mean "disable".  Definitely *don't* do that as it will make `undefined` and `null` silently assignable to all types.  You only get more errors *here* because by turning it off the compiler knows `a` can now be falsy and thus the RHS of `??` becomes reachable.

Technically `a ?? b` should be an unconditional error if the compiler knows `a` is always truthy.  It's clearly already doing that check anyway.

**Comment by Andarist** [comment]
> If you take the view that an expression of the form e ? b : c always includes reachable c even if e is truthy

My q here would be - why that view is taken by TS if other logical-like expressions don't take that view? I don't intend to argue with the choice. It's just something that I encountered in the past already and didn't know how to answer this for myself.
